File: src/algorithms/python/Minimum_Height_Trees.py
# -*- coding: utf-8 -*-


class Solution(object):
    def dfs(self, u, len):
        if len > self.longest_len:
            self.longest_p, self.longest_len = u, len
        for v in self.mp[u]:
            if not self.vis[v]:
                self.vis[v] = True
                self.dfs(v, len + 1)
                self.vis[v] = False

    def dfs_longest_path(self, u, len):
        if len > self.longest_len:
            self.longest_path, self.longest_len = self.tmp_path[:], len
        for v in self.mp[u]:
            if not self.vis[v]:
                self.vis[v] = True
                self.tmp_path.append(v)
                self.dfs_longest_path(v, len + 1)
                self.tmp_path.pop()
                self.vis[v] = False

    def findMinHeightTrees(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: List[int]
        """
        if n == 1:
            return [0]
        self.mp = dict([(i, list()) for i in range(n)])
        for u, v in edges:
            self.mp[u].append(v)
            self.mp[v].append(u)
        self.vis = [False for i in range(n)]
        self.longest_p, self.longest_len = -1, 1
        self.vis[0] = True
        self.dfs(0, 1)
        self.vis[0] = False
        print self.longest_p, self.longest_len
        self.longest_path, self.longest_len = [], 0
        self.vis[self.longest_p] = True
        self.tmp_path = [self.longest_p]
        self.dfs_longest_path(self.longest_p, 1)
        self.tmp_path.pop()
        self.vis[self.longest_p] = False
        if len(self.longest_path) % 2 == 1:
            return [self.longest_path[len(self.longest_path) // 2]]
        else:
            ret = [
                self.longest_path[len(self.longest_path) / 2 - 1],
                self.longest_path[len(self.longest_path) / 2]
            ]
            ret.sort()
            return ret


if __name__ == "__main__":
    solution = Solution()
    print solution.findMinHeightTrees(4, [[1, 0], [1, 2], [1, 3]])

File: src/algorithms/cpp/Minimum_Height_Trees.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
private:
    map <int, vector <int>> mp;
    vector <bool> vis;
    vector <int> tmp, longest_path;
    int longest_pt, max_len, n;

public:
    vector<int> findMinHeightTrees(int n, vector<pair<int, int>>& edges) {
        if (SZ(edges) == 0)     return vector <int> {0};
        this->longest_pt = 0;
        this->max_len = 0;
        this->n = n;
        for (int i = 0; i < SZ(edges); ++ i) {
            mp[edges[i].first].push_back(edges[i].second);
            mp[edges[i].second].push_back(edges[i].first);
        }
        for (int i = 0; i < n; ++ i)   vis.push_back(false);
        vis[0] = true;
        dfs(0, 1);
        for (int i = 0; i < n; ++ i)   vis[i] = false;
        this->max_len = 0;
        tmp.push_back(longest_pt);
        vis[longest_pt] = true;
        dfs_twice(longest_pt, 1);
        int mid = this->max_len / 2;
        if (this->max_len % 2 == 0)     return vector <int> {this->longest_path[mid - 1], this->longest_path[mid]};
        else                            return vector <int> {this->longest_path[mid]};
    }

private:
    void dfs(int u, int len) {
        if (len > max_len) {
            this->max_len = len;
            this->longest_pt = u;
        }
        for (int v: mp[u]) {
            if (!vis[v]) {
                vis[v] = true;
                this->dfs(v, len + 1);
            }
        }
    }

    void dfs_twice(int u, int len) {
        if (len > max_len) {
            this->max_len = len;
            this->longest_path = this->tmp;
        }
        for (int v: mp[u]) {
            if (!vis[v]) {
                vis[v] = true;
                this->tmp.push_back(v);
                this->dfs_twice(v, len + 1);
                this->tmp.pop_back();
            }
        }
    }
};

int main() {
    return 0;
}
